File: sage/services/snarc/detectors/reward.py
"""
Reward Estimator - Goal Relevance

Estimates how relevant sensor observations are to current goals.
Simple version: learns associations between sensor patterns and outcomes.
"""

import time as _time

import torch
import numpy as np
from typing import Any, Dict, List, Optional, Tuple
from collections import defaultdict

from sage.services.snarc.temporal import time_decay_weight, DEFAULT_HALF_LIVES


class RewardEstimator:
    """
    Estimate goal relevance of observations

    Learns associations: which sensor patterns precede positive outcomes?
    Uses simple similarity-based retrieval.
    """

    def __init__(self, memory_size: int = 500,
                 half_life: float = DEFAULT_HALF_LIVES['reward']):
        """
        Args:
            memory_size: How many outcome memories to keep
            half_life: Time-decay half-life in seconds for outcome weighting
        """
        self.memory_size = memory_size
        self.half_life = half_life

        # Memory: (sensor_output, reward_received, timestamp) tuples
        self.outcome_memory: Dict[str, List[Tuple[Any, float, float]]] = defaultdict(list)

        # Current goals (set by SAGE kernel)
        self.current_goals: List[str] = []

    def compute(self, sensor_output: Any, sensor_id: str) -> float:
        """
        Estimate reward relevance

        Args:
            sensor_output: Current sensor reading
            sensor_id: Unique sensor identifier

        Returns:
            Reward score (0.0-1.0)
                0.0 = irrelevant to goals
                1.0 = highly relevant to goals
        """
        # If no outcome history, neutral reward
        if sensor_id not in self.outcome_memory or not self.outcome_memory[sensor_id]:
            return 0.5

        # Find similar past observations
        similar_outcomes = self._find_similar_outcomes(sensor_output, sensor_id)

        if not similar_outcomes:
            return 0.5

        # Average reward of similar observations
        avg_reward = np.mean([reward for _, reward in similar_outcomes])

        # Normalize to 0-1 (assuming rewards in [-1, 1] range)
        normalized_reward = (avg_reward + 1.0) / 2.0

        return float(np.clip(normalized_reward, 0.0, 1.0))

    def update_from_outcome(self, sensor_output: Any, sensor_id: str, reward: float):
        """
        Learn from outcome

        Args:
            sensor_output: The sensor observation that was made
            sensor_id: Which sensor
            reward: Reward received after this observation (-1.0 to 1.0)
        """
        # Store outcome in memory
        if sensor_id not in self.outcome_memory:
            self.outcome_memory[sensor_id] = []

        memory = self.outcome_memory[sensor_id]
        memory.append((sensor_output, reward, _time.time()))

        # Limit memory size
        if len(memory) > self.memory_size:
            # Remove oldest entries
            self.outcome_memory[sensor_id] = memory[-self.memory_size:]

    def set_goals(self, goals: List[str]):
        """
        Set current goals

        Args:
            goals: List of goal descriptions (for future use)
        """
        self.current_goals = goals

    def _find_similar_outcomes(
        self,
        sensor_output: Any,
        sensor_id: str,
        top_k: int = 10
    ) -> List[Tuple[Any, float]]:
        """
        Find k most similar past observations

        Returns:
            List of (observation, reward) pairs
        """
        if sensor_id not in self.outcome_memory:
            return []

        memory = self.outcome_memory[sensor_id]
        now = _time.time()

        # Compute similarity to each past observation, weighted by recency
        similarities = []
        for entry in memory:
            past_obs, reward = entry[0], entry[1]
            ts = entry[2] if len(entry) > 2 else now  # backward compat
            raw_sim = self._compute_similarity(sensor_output, past_obs)
            recency = time_decay_weight(ts, now, self.half_life)
            effective_sim = raw_sim * recency
            similarities.append((effective_sim, past_obs, reward))

        # Sort by effective similarity (descending)
        similarities.sort(reverse=True, key=lambda x: x[0])

        # Return top k
        return [(obs, reward) for _, obs, reward in similarities[:top_k]]

    def _compute_similarity(self, current: Any, past: Any) -> float:
        """
        Compute similarity between observations

        Same as NoveltyDetector similarity computation
        """
        if isinstance(current, torch.Tensor) and isinstance(past, torch.Tensor):
            current_flat = current.flatten()
            past_flat = past.flatten()

            if current_flat.shape != past_flat.shape:
                return 0.0

            cos_sim = torch.nn.functional.cosine_similarity(
                current_flat.unsqueeze(0),
                past_flat.unsqueeze(0)
            ).item()

            return (cos_sim + 1.0) / 2.0

        elif isinstance(current, np.ndarray) and isinstance(past, np.ndarray):
            current_flat = current.flatten()
            past_flat = past.flatten()

            if current_flat.shape != past_flat.shape:
                return 0.0

            dot = np.dot(current_flat, past_flat)
            norm_product = np.linalg.norm(current_flat) * np.linalg.norm(past_flat)

            if norm_product == 0:
                return 0.0

            cos_sim = dot / norm_product
            return (cos_sim + 1.0) / 2.0

        elif isinstance(current, (int, float)) and isinstance(past, (int, float)):
            distance = abs(current - past)
            similarity = np.exp(-distance / 10.0)
            return float(similarity)

        else:
            return 0.5

    def get_statistics(self, sensor_id: str) -> Dict[str, Any]:
        """Get reward statistics for sensor"""
        if sensor_id not in self.outcome_memory or not self.outcome_memory[sensor_id]:
            return {}

        rewards = [entry[1] for entry in self.outcome_memory[sensor_id]]

        return {
            'num_outcomes': len(rewards),
            'mean_reward': float(np.mean(rewards)),
            'std_reward': float(np.std(rewards)),
            'min_reward': float(np.min(rewards)),
            'max_reward': float(np.max(rewards))
        }

    def reset_sensor(self, sensor_id: str):
        """Reset memory for specific sensor"""
        if sensor_id in self.outcome_memory:
            del self.outcome_memory[sensor_id]

    def reset_all(self):
        """Reset all memory"""
        self.outcome_memory.clear()

File: sage-rs/sage-lib/src/snarc/reward.rs
use std::collections::HashMap;

use crate::snarc::temporal::{default_half_life, time_decay_weight, now_secs};

struct OutcomeEntry {
    observation: f64,
    reward: f64,
    timestamp: f64,
}

/// Reward estimator — learns associations between observations and outcomes,
/// predicts goal relevance via similarity-weighted retrieval.
pub struct RewardEstimator {
    memory_size: usize,
    half_life: f64,
    outcome_memory: HashMap<String, Vec<OutcomeEntry>>,
}

impl RewardEstimator {
    pub fn new(memory_size: usize, half_life: Option<f64>) -> Self {
        Self {
            memory_size,
            half_life: half_life.unwrap_or_else(|| default_half_life("reward")),
            outcome_memory: HashMap::new(),
        }
    }

    pub fn with_defaults() -> Self {
        Self::new(500, None)
    }

    /// Estimate reward relevance. Returns 0.0 (irrelevant) to 1.0 (highly relevant).
    /// Returns 0.5 when there's no outcome history.
    pub fn compute(&self, observation: f64, sensor_id: &str) -> f64 {
        let entries = match self.outcome_memory.get(sensor_id) {
            Some(e) if !e.is_empty() => e,
            _ => return 0.5,
        };

        let similar = self.find_similar_outcomes(observation, entries, 10);
        if similar.is_empty() {
            return 0.5;
        }

        let avg_reward: f64 = similar.iter().map(|&r| r).sum::<f64>() / similar.len() as f64;
        // Normalize from [-1, 1] to [0, 1]
        ((avg_reward + 1.0) / 2.0).clamp(0.0, 1.0)
    }

    /// Record an outcome: the observation that was made, and the reward received.
    pub fn update_from_outcome(&mut self, observation: f64, sensor_id: &str, reward: f64) {
        let entries = self.outcome_memory
            .entry(sensor_id.to_string())
            .or_insert_with(Vec::new);

        entries.push(OutcomeEntry {
            observation,
            reward,
            timestamp: now_secs(),
        });

        if entries.len() > self.memory_size {
            let excess = entries.len() - self.memory_size;
            entries.drain(..excess);
        }
    }

    fn find_similar_outcomes(&self, observation: f64, entries: &[OutcomeEntry], top_k: usize) -> Vec<f64> {
        let now = now_secs();

        let mut scored: Vec<(f64, f64)> = entries
            .iter()
            .map(|e| {
                let sim = scalar_similarity(observation, e.observation);
                let recency = time_decay_weight(e.timestamp, now, self.half_life);
                (sim * recency, e.reward)
            })
            .collect();

        scored.sort_by(|a, b| b.0.partial_cmp(&a.0).unwrap_or(std::cmp::Ordering::Equal));
        scored.iter().take(top_k).map(|&(_, r)| r).collect()
    }

    pub fn reset_sensor(&mut self, sensor_id: &str) {
        self.outcome_memory.remove(sensor_id);
    }

    pub fn reset_all(&mut self) {
        self.outcome_memory.clear();
    }
}

fn scalar_similarity(a: f64, b: f64) -> f64 {
    (-(a - b).abs() / 10.0).exp()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn no_history_returns_neutral() {
        let est = RewardEstimator::with_defaults();
        assert_eq!(est.compute(5.0, "test"), 0.5);
    }

    #[test]
    fn positive_outcomes_predict_high() {
        let mut est = RewardEstimator::with_defaults();
        for _ in 0..20 {
            est.update_from_outcome(5.0, "test", 1.0);
        }
        let r = est.compute(5.0, "test");
        assert!(r > 0.9, "positive reward={r}, expected high");
    }

    #[test]
    fn negative_outcomes_predict_low() {
        let mut est = RewardEstimator::with_defaults();
        for _ in 0..20 {
            est.update_from_outcome(5.0, "test", -1.0);
        }
        let r = est.compute(5.0, "test");
        assert!(r < 0.1, "negative reward={r}, expected low");
    }

    #[test]
    fn mixed_outcomes_predict_by_similarity() {
        let mut est = RewardEstimator::with_defaults();
        // Observations near 5.0 → positive reward
        for _ in 0..10 {
            est.update_from_outcome(5.0, "test", 1.0);
        }
        // Observations near 50.0 → negative reward
        for _ in 0..10 {
            est.update_from_outcome(50.0, "test", -1.0);
        }
        // Query near 5.0 → should predict high reward
        let r_near_pos = est.compute(6.0, "test");
        // Query near 50.0 → should predict low reward
        let r_near_neg = est.compute(49.0, "test");
        assert!(r_near_pos > r_near_neg, "pos={r_near_pos}, neg={r_near_neg}");
    }

    #[test]
    fn memory_eviction() {
        let mut est = RewardEstimator::new(5, Some(600.0));
        for i in 0..10 {
            est.update_from_outcome(i as f64, "test", 0.5);
        }
        assert_eq!(est.outcome_memory.get("test").unwrap().len(), 5);
    }
}
